File: test/extensions/language/parameterized_types/java/parameterized_types/simple_param/SimpleParamTest.java
package parameterized_types.simple_param;

import static org.junit.jupiter.api.Assertions.*;

import java.io.IOException;

import org.junit.jupiter.api.Test;

import zserio.runtime.io.BitBuffer;
import zserio.runtime.io.BitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;
import zserio.runtime.io.SerializeUtil;

public class SimpleParamTest
{
    @Test
    public void parameterConstructor()
    {
        final Item item = new Item(LOWER_VERSION);
        assertFalse(item.isExtraParamUsed());
    }

    @Test
    public void bitStreamReaderConstructor() throws IOException
    {
        final long version = HIGHER_VERSION;
        final BitBuffer bitBuffer = writeItemToBitBuffer(version, ITEM_PARAM, ITEM_EXTRA_PARAM);
        final BitStreamReader stream = new ByteArrayBitStreamReader(bitBuffer);
        final Item item = new Item(stream, version);
        assertEquals(ITEM_PARAM, item.getParam());
        assertTrue(item.isExtraParamUsed());
        assertEquals(Long.valueOf(ITEM_EXTRA_PARAM), item.getExtraParam());
    }

    @Test
    public void fieldConstructor() throws IOException
    {
        final Item item = new Item(HIGHER_VERSION, ITEM_PARAM, (long)ITEM_EXTRA_PARAM);
        assertEquals(ITEM_PARAM, item.getParam());
        assertTrue(item.isExtraParamUsed());
        assertEquals(Long.valueOf(ITEM_EXTRA_PARAM), item.getExtraParam());
    }

    @Test
    public void bitSizeOf()
    {
        final Item item1 = new Item(LOWER_VERSION, ITEM_PARAM, Long.valueOf(ITEM_EXTRA_PARAM));
        assertEquals(ITEM_BIT_SIZE_WITHOUT_OPTIONAL, item1.bitSizeOf());

        final Item item2 = new Item(HIGHER_VERSION, ITEM_PARAM, Long.valueOf(ITEM_EXTRA_PARAM));
        assertEquals(ITEM_BIT_SIZE_WITH_OPTIONAL, item2.bitSizeOf());
    }

    @Test
    public void equals()
    {
        final Item item1 = new Item(LOWER_VERSION);
        final Item item2 = new Item(LOWER_VERSION);
        assertTrue(item1.equals(item2));

        final Item item3 = new Item(HIGHER_VERSION);
        assertFalse(item2.equals(item3));
    }

    @Test
    public void hashCodeMethod()
    {
        final Item item1 = new Item(LOWER_VERSION);
        final Item item2 = new Item(LOWER_VERSION);
        assertEquals(item1.hashCode(), item2.hashCode());

        final Item item3 = new Item(HIGHER_VERSION);
        assertTrue(item2.hashCode() != item3.hashCode());
    }

    @Test
    public void initializeOffsets()
    {
        final Item item1 = new Item(LOWER_VERSION, ITEM_PARAM, Long.valueOf(ITEM_EXTRA_PARAM));
        final int bitPosition = 1;
        assertEquals(bitPosition + ITEM_BIT_SIZE_WITHOUT_OPTIONAL, item1.initializeOffsets(bitPosition));

        final Item item2 = new Item(HIGHER_VERSION, ITEM_PARAM, Long.valueOf(ITEM_EXTRA_PARAM));
        assertEquals(bitPosition + ITEM_BIT_SIZE_WITH_OPTIONAL, item2.initializeOffsets(bitPosition));
    }

    @Test
    public void writeRead() throws IOException
    {
        final long version1 = LOWER_VERSION;
        final Item item1 = new Item(version1);
        item1.setParam(ITEM_PARAM);
        final BitBuffer bitBuffer1 = SerializeUtil.serialize(item1);
        checkItemInBitBuffer(bitBuffer1, item1, version1);
        Item readItem1 = SerializeUtil.deserialize(Item.class, bitBuffer1, version1);
        assertEquals(item1, readItem1);

        final long version2 = HIGHER_VERSION;
        final Item item2 = new Item(version2, ITEM_PARAM, (long)ITEM_EXTRA_PARAM);
        final BitBuffer bitBuffer2 = SerializeUtil.serialize(item2);
        checkItemInBitBuffer(bitBuffer2, item2, version2);
        Item readItem2 = SerializeUtil.deserialize(Item.class, bitBuffer2, version2);
        assertEquals(item2, readItem2);
    }

    private BitBuffer writeItemToBitBuffer(long version, short param, int extraParam) throws IOException
    {
        try (final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter())
        {
            writer.writeBits(param, 16);
            if (version >= HIGHER_VERSION)
                writer.writeBits(extraParam, 32);
            return new BitBuffer(writer.toByteArray(), writer.getBitPosition());
        }
    }

    private void checkItemInBitBuffer(BitBuffer bitBuffer, Item item, long version) throws IOException
    {
        try (final ByteArrayBitStreamReader reader = new ByteArrayBitStreamReader(bitBuffer))
        {
            assertEquals(item.getParam(), reader.readBits(16));
            if (version >= HIGHER_VERSION)
                assertEquals(item.getExtraParam(), Long.valueOf(reader.readBits(32)));
        }
    }

    private static final long LOWER_VERSION = 9;
    private static final long HIGHER_VERSION = 10;

    private static final short ITEM_PARAM = 0xAA;
    private static final int ITEM_EXTRA_PARAM = 0xBB;

    private static final int ITEM_BIT_SIZE_WITHOUT_OPTIONAL = 16;
    private static final int ITEM_BIT_SIZE_WITH_OPTIONAL = 16 + 32;
}

File: test/extensions/language/parameterized_types/cpp/SimpleParamTest.cpp
#include "gtest/gtest.h"
#include "parameterized_types/simple_param/Item.h"
#include "zserio/BitStreamReader.h"
#include "zserio/BitStreamWriter.h"
#include "zserio/CppRuntimeException.h"

namespace parameterized_types
{
namespace simple_param
{

class SimpleParamTest : public ::testing::Test
{
protected:
    void writeItemToByteArray(
            zserio::BitStreamWriter& writer, uint32_t version, uint16_t param, uint32_t extraParam)
    {
        writer.writeBits(param, 16);
        if (version >= HIGHER_VERSION)
        {
            writer.writeBits(extraParam, 32);
        }
    }

    void checkItemInBitStream(zserio::BitStreamReader& reader, uint32_t version, const Item& item)
    {
        ASSERT_EQ(item.getParam(), reader.readBits(16));
        if (version >= HIGHER_VERSION)
        {
            ASSERT_EQ(item.getExtraParam(), reader.readBits(32));
        }
    }

    static const uint32_t LOWER_VERSION;
    static const uint32_t HIGHER_VERSION;

    static const uint16_t ITEM_PARAM;
    static const uint32_t ITEM_EXTRA_PARAM;

    static const size_t ITEM_BIT_SIZE_WITHOUT_OPTIONAL;
    static const size_t ITEM_BIT_SIZE_WITH_OPTIONAL;

    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
};

const uint32_t SimpleParamTest::LOWER_VERSION = 9;
const uint32_t SimpleParamTest::HIGHER_VERSION = 10;

const uint16_t SimpleParamTest::ITEM_PARAM = 0xAA;
const uint32_t SimpleParamTest::ITEM_EXTRA_PARAM = 0xBB;

const size_t SimpleParamTest::ITEM_BIT_SIZE_WITHOUT_OPTIONAL = 16;
const size_t SimpleParamTest::ITEM_BIT_SIZE_WITH_OPTIONAL = 16 + 32;

TEST_F(SimpleParamTest, emptyConstructor)
{
    const uint16_t version = LOWER_VERSION;

    {
        Item item;
        item.initialize(version);
        ASSERT_EQ(version, item.getVersion());
        ASSERT_FALSE(item.isExtraParamUsed());
    }

    {
        Item item = {};
        item.initialize(version);
        ASSERT_EQ(version, item.getVersion());
        ASSERT_EQ(0, item.getParam());
        ASSERT_FALSE(item.isExtraParamUsed());
    }
}

TEST_F(SimpleParamTest, fieldConstructor)
{
    {
        Item item(ITEM_PARAM, zserio::NullOpt);
        ASSERT_FALSE(item.isInitialized());
        item.initialize(LOWER_VERSION);
        ASSERT_FALSE(item.isExtraParamUsed());
    }
    {
        Item item(ITEM_PARAM, ITEM_EXTRA_PARAM);
        ASSERT_FALSE(item.isInitialized());
        item.initialize(HIGHER_VERSION);
        ASSERT_TRUE(item.isExtraParamUsed());
        ASSERT_EQ(ITEM_EXTRA_PARAM, item.getExtraParam());
    }

    {
        Item item({}, {});
        ASSERT_FALSE(item.isInitialized());
        item.initialize(HIGHER_VERSION);
        ASSERT_TRUE(item.isExtraParamUsed());
        ASSERT_EQ(0, item.getParam());
        ASSERT_EQ(0, item.getExtraParam());
    }
}

TEST_F(SimpleParamTest, bitStreamReaderConstructor)
{
    const uint32_t version = HIGHER_VERSION;
    zserio::BitStreamWriter writer(bitBuffer);
    const uint16_t itemParam = ITEM_PARAM;
    const uint32_t itemExtraParam = ITEM_EXTRA_PARAM;
    writeItemToByteArray(writer, version, itemParam, itemExtraParam);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    Item item(reader, version);
    ASSERT_EQ(version, item.getVersion());
    ASSERT_EQ(itemParam, item.getParam());
    ASSERT_TRUE(item.isExtraParamUsed());
    ASSERT_EQ(itemExtraParam, item.getExtraParam());
}

TEST_F(SimpleParamTest, copyConstructor)
{
    Item item;
    item.setParam(ITEM_PARAM);
    item.setExtraParam(ITEM_EXTRA_PARAM);

    const Item itemCopy1(item);
    ASSERT_THROW(itemCopy1.getVersion(), zserio::CppRuntimeException);

    item.initialize(HIGHER_VERSION);
    const Item itemCopy2(item);
    ASSERT_EQ(item.getVersion(), itemCopy2.getVersion());
}

TEST_F(SimpleParamTest, operatorAssignment)
{
    Item item;
    item.setParam(ITEM_PARAM);
    item.setExtraParam(ITEM_EXTRA_PARAM);

    const Item itemCopy1 = item;
    ASSERT_THROW(itemCopy1.getVersion(), zserio::CppRuntimeException);

    item.initialize(HIGHER_VERSION);
    const Item itemCopy2 = item;
    ASSERT_EQ(item.getVersion(), itemCopy2.getVersion());
}

TEST_F(SimpleParamTest, moveConstructor)
{
    Item item(ITEM_PARAM, zserio::NullOpt);
    item.initialize(LOWER_VERSION);

    Item movedItem(std::move(item));
    ASSERT_EQ(ITEM_PARAM, movedItem.getParam());
    ASSERT_FALSE(movedItem.isExtraParamUsed());
}

TEST_F(SimpleParamTest, moveAssignmentOperator)
{
    Item item(ITEM_PARAM, ITEM_EXTRA_PARAM);
    item.initialize(HIGHER_VERSION);

    Item movedItem;
    movedItem = std::move(item);
    ASSERT_EQ(ITEM_PARAM, movedItem.getParam());
    ASSERT_TRUE(movedItem.isExtraParamUsed());
    ASSERT_EQ(ITEM_EXTRA_PARAM, movedItem.getExtraParam());
}

TEST_F(SimpleParamTest, propagateAllocatorCopyConstructor)
{
    Item item;
    item.setParam(ITEM_PARAM);
    item.setExtraParam(ITEM_EXTRA_PARAM);

    const Item itemCopy1(zserio::PropagateAllocator, item, Item::allocator_type());
    ASSERT_THROW(itemCopy1.getVersion(), zserio::CppRuntimeException);

    item.initialize(HIGHER_VERSION);
    const Item itemCopy2(zserio::PropagateAllocator, item, Item::allocator_type());
    ASSERT_EQ(item.getVersion(), itemCopy2.getVersion());
}

TEST_F(SimpleParamTest, bitSizeOf)
{
    Item item1;
    item1.initialize(LOWER_VERSION);
    item1.setParam(ITEM_PARAM);
    const size_t expectedBitSizeWithoutOptional = ITEM_BIT_SIZE_WITHOUT_OPTIONAL;
    ASSERT_EQ(expectedBitSizeWithoutOptional, item1.bitSizeOf());

    Item item2;
    item2.initialize(HIGHER_VERSION);
    item2.setParam(ITEM_PARAM);
    item2.setExtraParam(ITEM_EXTRA_PARAM);
    const size_t expectedBitSizeWithOptional = ITEM_BIT_SIZE_WITH_OPTIONAL;
    ASSERT_EQ(expectedBitSizeWithOptional, item2.bitSizeOf());
}

TEST_F(SimpleParamTest, initializeOffsets)
{
    Item item1;
    item1.initialize(LOWER_VERSION);
    item1.setParam(ITEM_PARAM);
    const size_t bitPosition = 1;
    const size_t expectedBitPositionWithoutOptional = bitPosition + ITEM_BIT_SIZE_WITHOUT_OPTIONAL;
    ASSERT_EQ(expectedBitPositionWithoutOptional, item1.initializeOffsets(bitPosition));

    Item item2;
    item2.initialize(HIGHER_VERSION);
    item2.setParam(ITEM_PARAM);
    item2.setExtraParam(ITEM_EXTRA_PARAM);
    const size_t expectedBitPositionWithOptional = bitPosition + ITEM_BIT_SIZE_WITH_OPTIONAL;
    ASSERT_EQ(expectedBitPositionWithOptional, item2.initializeOffsets(bitPosition));
}

TEST_F(SimpleParamTest, operatorEquality)
{
    Item item1(ITEM_PARAM, ITEM_EXTRA_PARAM);
    item1.initialize(LOWER_VERSION);
    Item item2(ITEM_PARAM, ITEM_EXTRA_PARAM);
    item2.initialize(LOWER_VERSION);
    ASSERT_TRUE(item1 == item2);

    Item item3(ITEM_PARAM, ITEM_EXTRA_PARAM);
    item3.initialize(HIGHER_VERSION);

    ASSERT_FALSE(item2 == item3);
}

TEST_F(SimpleParamTest, operatorLessThan)
{
    Item item1(ITEM_PARAM, ITEM_EXTRA_PARAM);
    item1.initialize(LOWER_VERSION);
    Item item2(ITEM_PARAM, ITEM_EXTRA_PARAM);
    item2.initialize(LOWER_VERSION);
    ASSERT_FALSE(item1 < item2);
    ASSERT_FALSE(item2 < item1);

    Item item3(ITEM_PARAM, ITEM_EXTRA_PARAM);
    item3.initialize(HIGHER_VERSION);
    ASSERT_TRUE(item2 < item3);
    ASSERT_FALSE(item3 < item2);
}

TEST_F(SimpleParamTest, hashCode)
{
    Item item1(ITEM_PARAM, ITEM_EXTRA_PARAM);
    item1.initialize(LOWER_VERSION);
    Item item2(ITEM_PARAM, ITEM_EXTRA_PARAM);
    item2.initialize(LOWER_VERSION);
    ASSERT_EQ(item1.hashCode(), item2.hashCode());

    Item item3(ITEM_PARAM, ITEM_EXTRA_PARAM);
    item3.initialize(HIGHER_VERSION);
    ASSERT_TRUE(item2.hashCode() != item3.hashCode());
}

TEST_F(SimpleParamTest, write)
{
    const uint32_t version = HIGHER_VERSION;
    Item item;
    item.initialize(version);
    item.setParam(ITEM_PARAM);
    item.setExtraParam(ITEM_EXTRA_PARAM);

    zserio::BitStreamWriter writer(bitBuffer);
    item.write(writer);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    checkItemInBitStream(reader, version, item);
    reader.setBitPosition(0);

    const Item readItem(reader, version);
    ASSERT_EQ(item, readItem);
}

} // namespace simple_param
} // namespace parameterized_types
